### PR TITLE
Log predictions during training to W&B Tables

### PR DESCRIPTION
**Pull request by manangoel99**
This PR adds a lot of functionality to the the current W&B 🪄🐝  Integration.
==================
The current W&B integration logs the training and validation metrics as well as the checkpoints to the user's wandb dashboard. This PR will allow logging of the predicted bounding boxes on the validation set to be logged to the dashboard as well!

### Changes
- The images in the validation set are automatically logged to the wandb dashboard. To turn off logging add the flag `wandb-num_eval_images 0` to the command.
- An added argument to the evaluator `return_outputs=False`. This returns the outputs for each image in the validation set as a dict.  Formatted as 
```
{
    img_id: {
         bboxes: <list of bounding boxes>
         scores: <list of scores for each bounding box>
         categories: <category for each bounding box>
    }
}
```
- This dict of predictions is then sent to `WandbLogger.log_images` which logs `num_eval_images` predictions on the validation set to the dashboard along with the average score for each class.
- The logged checkpoints also possess metadata including optimizer state at the checkpoint and corresponding AP on the validation set.
Note: The `val_dataset` argument in `WandbLogger.__init__` is not open to the CLI.

### Usage
The WandbLogger will by default log 100 images from the validation set to the dashboard. To change any default arguments, use the following reference
```
            python tools/train.py .... --logger wandb \
                wandb-project <project-name> \
                wandb-name <run-name> \
                wandb-id <run-id> \
                wandb-save_dir <save-dir> \
                wandb-num_eval_imges <num-images> \
                wandb-log_checkpoints <bool>
```
[Example Workspace](https://wandb.ai/manan-goel/yolox-nano/runs/3ntph3k)
More information is available [here](https://docs.wandb.ai/guides/integrations/other/yolox) (WIP)

Sneak Peek 🔥🔥🔥
===============

### Log Training, Validation and System Metrics
Screenshot 2022-03-16 at 4 13 25 PM
Screenshot 2022-03-16 at 4 13 43 PM

### Log Model Checkpoints with Metadata

https://user-images.githubusercontent.com/32260633/158601288-d0c60c18-0f13-4a6d-bac0-1fdec133884e.mp4

### Log Predicted bounding boxes and class scores

https://user-images.githubusercontent.com/32260633/158601599-e4167732-5e3c-4dd4-bc8c-0f0e9e537623.mp4

Looking forward to the reviews!

### CODE REVIEW

**Comment by FateScript** [comment]
Thanks for your contribtion, it's a PR with lots of features, we need more time to review this.

**Comment by manangoel99** [comment]
Hi @FateScript Any updates on the review?

**Comment by FateScript** [comment]
> Hi @FateScript Any updates on the review?

Sorry that I'm busy with my work these days. I will review your code as soon as possible.

**Comment by manangoel99** [comment]
Not a problem! Please take your time. Thank you for getting back to me on this

File: yolox/data/datasets/coco.py [hunk 1]
```diff
@@ -65,8 +65,8 @@ def __init__(
         remove_useless_info(self.coco)
         self.ids = self.coco.getImgIds()
         self.class_ids = sorted(self.coco.getCatIds())
-        cats = self.coco.loadCats(self.coco.getCatIds())
-        self._classes = tuple([c["name"] for c in cats])
+        self.cats = self.coco.loadCats(self.coco.getCatIds())
```
Comment (FateScript) [inline yolox/data/datasets/coco.py:68]: why `cats` becomes attribute here?

File: README.md [hunk 2]
```diff
@@ -148,6 +148,19 @@ On the second machine, run
 python tools/train.py -n yolox-s -b 128 --dist-url tcp://123.123.123.123:12312 --num-machines 2 --machine-rank 1
 ```
 
+**Logging to Weights & Biases**
```
[the remark this answers is not in the document]
Comment (manangoel99) [reply README.md:151]: Will move all the W&B related documentation to one file

File: yolox/core/trainer.py [hunk 3]
```diff
@@ -336,16 +339,17 @@ def evaluate_and_save_model(self):
...
         synchronize()
 
-        self.save_ckpt("last_epoch", update_best_ckpt)
+        self.save_ckpt("last_epoch", update_best_ckpt, ap=ap50_95)
         if self.save_history_ckpt:
-            self.save_ckpt(f"epoch_{self.epoch + 1}")
+            self.save_ckpt(f"epoch_{self.epoch + 1}", ap=ap50_95)
 
-    def save_ckpt(self, ckpt_name, update_best_ckpt=False):
+    def save_ckpt(self, ckpt_name, update_best_ckpt=False, ap=None):
```
[the remark this answers is not in the document]
Comment (manangoel99) [reply yolox/core/trainer.py:352]: The reason for adding `ap` as an argument was to be able to log it as metadata along with the model parameters. Making `current_ap` an attribute would remove the requirement for changing the function signature so that would be a better case. Will update this.

File: yolox/evaluators/coco_evaluator.py [hunk 4]
```diff
@@ -120,6 +121,7 @@ def evaluate(
         trt_file=None,
         decoder=None,
         test_size=None,
+        return_outputs=False
```
[the remark this answers is not in the document]
Comment (manangoel99) [reply yolox/evaluators/coco_evaluator.py:124]: The reason behind this is that when the evaluator finds the outputs for the validation samples, all the metrics are calculated and returned and the outputs are not. So in order to be able to log the visualizations on the dashboard, we need the output bounding boxes as well.
Hence, I added a `return_outputs` flag to the function to return outputs and the default value is False so that it doesn't break if the function is called anywhere else.

File: yolox/evaluators/coco_evaluator.py [hunk 5]
```diff
@@ -226,6 +238,20 @@ def convert_to_coco_format(self, outputs, info_imgs, ids):
                     "segmentation": [],
                 }  # COCO json format
                 data_list.append(pred_data)
+
+            image_wise_data.update({
```
[the remark this answers is not in the document]
Comment (manangoel99) [reply yolox/evaluators/coco_evaluator.py:242]: The logic here is to create a dictionary with the `image_id` as the key and the value contains

- The coordinates of all the bounding boxes in the image
- The confidence score for each bounding box
- The label for each bounding box.

These outputs are then returned if `return_outputs=True`.

File: yolox/utils/logger.py [hunk 6]
```diff
@@ -125,7 +133,24 @@ def __init__(self,
             entity (str): wandb entity name.
             save_dir (str): save directory.
             config (dict): config dict.
+            val_dataset (Dataset): validation dataset.
+            num_eval_images (int): number of images from the validation set to log.
+            log_checkpoints (bool): log checkpoints
             **kwargs: other kwargs.
+
+        Usage:
```
[the remark this answers is not in the document]
Comment (manangoel99) [reply yolox/utils/logger.py:141]: Will move it accordingly

File: yolox/utils/logger.py [hunk 7]
```diff
@@ -188,21 +254,98 @@ def log_metrics(self, metrics, step=None):
...
 
         if step is not None:
-            self.run.log(metrics, step=step)
+            metrics.update({"train/step": step})
+            self.run.log(metrics)
         else:
             self.run.log(metrics)
 
-    def save_checkpoint(self, save_dir, model_name, is_best):
+    def log_images(self, predictions):
```
[the remark this answers is not in the document]
Comment (manangoel99) [reply yolox/utils/logger.py:262]: Will fix it

File: yolox/data/datasets/coco.py (hunk 1 above)
Comment (manangoel99) [reply yolox/data/datasets/coco.py:68]: `cats` was changed to an attribute so that the category names can be accessed by other functions.

**Comment by manangoel99** [comment]
@FateScript Thank you for the thorough review! I will make the suggested changes and please let me know if you have more questions about the proposed changes.

**Comment by manangoel99** [comment]
Hi @FateScript ! Please have a look whenever you get the chance

**Comment by dnth** [comment]
This is an exciting update! Is it going to be merged soon?

**Comment by manangoel99** [comment]
@FateScript would love some feedback on the PR whenever you have some time

**Comment by FateScript** [comment]
> @FateScript would love some feedback on the PR whenever you have some time

I will review this PR today.

**Comment by FateScript** [comment]
Any updates ? @manangoel99

**Comment by manangoel99** [comment]
Hi @FateScript apologies for the delay. I was travelling.
I've rebased with master and fixed a minor bug that came up.
Please have a look

File: yolox/evaluators/coco_evaluator.py (hunk 5 above)
Comment (totti0223) [inline yolox/evaluators/coco_evaluator.py:242]: image_wise_data requires xyxy(xmin,ymin,xmax,ymax) format, but is currently feeding xywh(xmin,ymin,width,height) format due to `bboxes = xyxy2xywh(bboxes)` for coco.

therefore 

`image_wise_data.update{.....` 
should be should be moved to between the following lines
```
scores = output[:, 4] * output[:, 5]
for ind in range(bboxes.shape[0]):
```
while 
`bboxes = xyxy2xywh(bboxes)`
should be moved after `image_wise_data.update{.....`  and before 'for ind in range(bboxes.shape[0]):'


An alternative can be modifying logger.py arround the following code to reconvert xywh to xyxy.
will leave the decision to modify which depending on code clarity.

`                for i in range(len(prediction["bboxes"])):
                    bbox = prediction["bboxes"][i]
                    x0 = bbox[0]
                    y0 = bbox[1]
                    x1 = bbox[2]
                    y1 = bbox[3]
                    box = {
                        "position": {
                            "minX": min(x0, x1),
                            "minY": min(y0, y1),
                            "maxX": max(x0, x1),
                            "maxY": max(y0, y1)`


Anyway, the above modification solved the visualization bug in wandb browser.

Comment (manangoel99) [reply yolox/evaluators/coco_evaluator.py:230]: Thanks for the great pointer. I added this and it works like a charm!

**Comment by manangoel99** [comment]
@FateScript Would love some inputs whenever possible

**Comment by FateScript** [comment]
@manangoel99 I would love to merge this and I need some feedback from other maintainer. If this PR is not merged in 3 days, Plz let me know ! Thanks for your contribution.

**Review by FateScript — APPROVED** [review]

**Comment by FateScript** [comment]
Functionality looks good to me.
Some code in this PR  might create complexity. Refactor is needed in the future.
